// @serenity-js/core/lib/screenplay/interactions/TakeNote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TakeNote = void 0;
const io_1 = require("../../io");
const abilities_1 = require("../abilities");
const Interaction_1 = require("../Interaction");
/**
 * @desc
 *  Enables the {@link Actor} to remember an answer to a given {@link Question},
 *  and recall it later.
 *
 * @example <caption>Using default subject name based on the name of the question</caption>
 *  import { actorCalled, Note, TakeNote, TakeNotes } from '@serenity-js/core'
 *  import { BrowseTheWeb, Target, Text } from '@serenity-js/protractor'
 *  import { by, protractor } from 'protractor';
 *
 *  class Vouchers {
 *      static code             = Target.the('voucher code').located(by.id('voucher'));
 *      static appliedVoucher   = Target.the('voucher code').located(by.id('applied-voucher'));
 *  }
 *
 *  const actor = actorCalled('Noah').whoCan(
 *      TakeNotes.usingAnEmptyNotepad(),
 *      BrowseTheWeb.using(protractor.browser),
 *  );
 *
 *  actor.attemptsTo(
 *      TakeNote.of(Text.of(Vouchers.code)),
 *      // ... add the product to a basket, go to checkout, etc.
 *      Ensure.that(Text.of(Vouchers.appliedVoucher), equals(Note.of(Text.of(Vouchers.code)))),
 *  );
 *
 * @example <caption>Using custom subject name</caption>
 *  actor.attemptsTo(
 *      TakeNote.of(Text.of(Vouchers.code)).as('voucher code'),
 *      // ... add the product to a basket, go to checkout, etc.
 *      Ensure.that(Text.of(Vouchers.appliedVoucher), equals(Note.of('voucher code'))),
 *  );
 *
 * @see {@link Note}
 * @see {@link TakeNotes}
 *
 * @extends {Interaction}
 */
class TakeNote extends Interaction_1.Interaction {
    /**
     * @param {Question<Promise<Answer>> | Question<Answer>} question
     * @param {string} subject
     */
    constructor(question, subject = question.toString()) {
        super();
        this.question = question;
        this.subject = subject;
    }
    /**
     * @desc
     *  Instructs the {@link Actor} to remember the answer to a given question
     *
     * @param {Question<Promise<A>> | Question<A>} question
     *
     * @returns {TakeNote<A>}
     */
    static of(question) {
        return new TakeNote(question);
    }
    /**
     * @desc
     *  Sets a custom subject name to remember the answer as.
     *
     * @param {string} subject
     * @returns {TakeNote<Answer>}
     */
    as(subject) {
        return new TakeNote(this.question, subject);
    }
    /**
     * @desc
     *  Makes the provided {@link Actor}
     *  perform this {@link Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     * @returns {Promise<void>}
     *
     * @see {@link Actor}
     * @see {@link UsesAbilities}
     * @see {@link AnswersQuestions}
     */
    performAs(actor) {
        return actor.answer(this.question)
            .then(answer => abilities_1.TakeNotes.as(actor).record(this.subject, answer));
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor takes note of ${this.subject}`;
    }
}
exports.TakeNote = TakeNote;
