// @serenity-js/protractor/lib/screenplay/questions/Attribute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Attribute = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
const withAnswerOf_1 = require("../withAnswerOf");
// todo: it might be better to swap the order of arguments
//  - Attribute.called('href').of(link) to make it work with ArrayListFilter
class Attribute extends core_1.Question {
    constructor(target, name) {
        super(io_1.formatted `the value of the ${name} attribute of ${target}`);
        this.target = target;
        this.name = name;
    }
    /**
     * @param {Question<ElementFinder> | ElementFinder} target
     * @returns {AttributeBuilder}
     */
    static of(target) {
        return {
            called: (name) => new Attribute(target, name),
        };
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  answer this {@link @serenity-js/core/lib/screenplay~Question}.
     *
     * @param {AnswersQuestions & UsesAbilities} actor
     * @returns {Promise<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     */
    answeredBy(actor) {
        return actor.answer(this.name)
            .then(name => withAnswerOf_1.withAnswerOf(actor, this.target, elf => elf.getAttribute(name)));
    }
}
exports.Attribute = Attribute;
