// @serenity-js/protractor/lib/screenplay/questions/targets/TargetNestedElements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TargetNestedElements = void 0;
const core_1 = require("@serenity-js/core");
const withAnswerOf_1 = require("../../withAnswerOf");
const override_1 = require("./override");
const lists_1 = require("../lists");
/**
 * @desc
 *  Locates a group of {@link WebElement}s located within another {@link WebElement}.
 *
 *  Instead of using this class directly, please use {@link Target.all} and {@link TargetElements#of} instead.
 *
 * @public
 * @see {@link Target}
 *
 * @extends {@serenity-js/core/lib/screenplay~Question}
 * @implements {@serenity-js/core/lib/screenplay/questions~MetaQuestion}
 */
class TargetNestedElements extends core_1.Question {
    /**
     * @desc
     *
     * @param {Question<ElementFinder> | ElementFinder} parent
     * @param {Question<ElementArrayFinder> | ElementArrayFinder} children
     */
    constructor(parent, children) {
        super(`${children.toString()} of ${parent}`);
        this.parent = parent;
        this.children = children;
        this.list = new core_1.List(new lists_1.ElementArrayFinderListAdapter(this));
    }
    /**
     * @desc
     *  Retrieves a group of {@link WebElement}s located by `locator`,
     *  resolved in the context of a `parent` {@link WebElement}.
     *
     * @param {Question<ElementFinder> | ElementFinder} parent
     * @returns {TargetNestedElements}
     *
     * @see {@link Target}
     */
    of(parent) {
        return new TargetNestedElements(parent, this);
    }
    /**
     * @desc
     *  Returns the number of {@link ElementFinder}s matched by the `locator`
     *
     * @returns {@serenity-js/core/lib/screenplay~Question<Promise<number>>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~List}
     */
    count() {
        return this.list.count();
    }
    /**
     * @desc
     *  Returns the first of {@link ElementFinder}s matched by the `locator`
     *
     * @returns {@serenity-js/core/lib/screenplay~Question<ElementFinder>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~List}
     */
    first() {
        return this.list.first();
    }
    /**
     * @desc
     *  Returns the last of {@link ElementFinder}s matched by the `locator`
     *
     * @returns {@serenity-js/core/lib/screenplay~Question<ElementFinder>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~List}
     */
    last() {
        return this.list.last();
    }
    /**
     * @desc
     *  Returns an {@link ElementFinder} at `index` for `locator`
     *
     * @param {number} index
     *
     * @returns {@serenity-js/core/lib/screenplay~Question<ElementFinder>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~List}
     */
    get(index) {
        return this.list.get(index);
    }
    /**
     * @desc
     *  Filters the list of {@link ElementFinder}s matched by `locator` to those that meet the additional {@link @serenity-js/core/lib/screenplay/questions~Expectation}s.
     *
     * @param {@serenity-js/core/lib/screenplay/questions~MetaQuestion<ElementFinder, Promise<Answer_Type> | Answer_Type>} question
     * @param {@serenity-js/core/lib/screenplay/questions~Expectation<any, Answer_type>} expectation
     *
     * @returns {@serenity-js/core/lib/screenplay/questions~List<ElementArrayFinderListAdapter, ElementFinder, ElementArrayFinder>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/questions~List}
     */
    where(question, expectation) {
        return this.list.where(question, expectation);
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  answer this {@link @serenity-js/core/lib/screenplay~Question}.
     *
     * @param {AnswersQuestions & UsesAbilities} actor
     * @returns {Promise<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     */
    answeredBy(actor) {
        return withAnswerOf_1.withAnswerOf(actor, this.parent, parent => withAnswerOf_1.withAnswerOf(actor, this.children, children => override_1.override(parent.all(children.locator()), 'toString', this.toString.bind(this))));
    }
}
exports.TargetNestedElements = TargetNestedElements;
