// @serenity-js/protractor/lib/screenplay/interactions/Dismiss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dismiss = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
/**
 * @desc
 *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
 *  dismiss a {@link ModalDialog}.
 *
 * @example <caption>Example widget</caption>
 *  <button
 *      data-test="trigger"
 *      onclick="alert('hello!')">Trigger Alert</button>
 *
 * @example <caption>Lean Page Object describing the widget</caption>
 *  import { Target } from '@serenity-js/protractor';
 *  import { by } from 'protractor';
 *
 *  class Widget {
 *      static trigger = Target.the('trigger button')
 *          .located(by.css('[data-test="trigger"]'));
 *  }
 *
 * @example <caption>Dismissing a modal dialog window</caption>
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Click, Dismiss, ModalDialog } from '@serenity-js/protractor';
 *  import { protractor } from 'protractor';
 *
 *  actorCalled('Nick')
 *      .whoCan(BrowseTheWeb.using(protractor.browser))
 *      .attemptsTo(
 *          Click.on(Widget.trigger),
 *          Dismiss.the(ModalDialog.window()),
 *      );
 *
 * @extends {@serenity-js/core/lib/screenplay~Interaction}
 *
 * @see {@link Accept}
 * @see {@link ModalDialog}
 */
class Dismiss extends core_1.Interaction {
    /**
     * @param {@serenity-js/core/lib/screenplay~Question<AlertPromise> | AlertPromise} modalDialogWindow
     *  The modal dialog window to dismiss
     *
     * @see {@link ModalDialog.window}
     */
    constructor(modalDialogWindow) {
        super();
        this.modalDialogWindow = modalDialogWindow;
    }
    /**
     * @desc
     *  Instantiates this {@link @serenity-js/core/lib/screenplay~Interaction}
     *  with a {@link ModalDialog.window} the {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  will dismiss.
     *
     * @param {@serenity-js/core/lib/screenplay~Question<AlertPromise> | AlertPromise} modalDialogWindow
     *  The modal dialog window to dismiss
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link ModalDialog.window}
     * @see {@link @serenity-js/core/lib/screenplay~Question}
     */
    static the(modalDialogWindow) {
        return new Dismiss(modalDialogWindow);
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return actor.answer(this.modalDialogWindow)
            .then(alert => alert.dismiss());
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor dismisses ${this.modalDialogWindow}`;
    }
}
exports.Dismiss = Dismiss;
