// @serenity-js/protractor/lib/screenplay/interactions/Navigate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Navigate = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
const promiseOf_1 = require("../../promiseOf");
const abilities_1 = require("../abilities");
/**
 * @desc
 *  Allows the {@link @serenity-js/core/lib/screenplay/actor~Actor} to navigate to a specific destination,
 *  as well as back and forth in the browser history, or reload the current page.
 */
class Navigate {
    /**
     * @desc
     *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
     *  navigate to a given URL.
     *
     *  The URL can be:
     *  - absolute, i.e. `https://example.org/search`
     *  - relative, i.e. `/search`
     *
     *  If the URL is relative, Protractor will append it to `baseUrl` configured in [`protractor.conf.js`](https://github.com/angular/protractor/blob/master/lib/config.ts).
     *
     * @example <caption>protractor.conf.js</caption>
     *  exports.config = {
     *      baseUrl: 'https://example.org',
     *      // ...
     *  }
     *
     * @example <caption>Navigate to path relative to baseUrl</caption>
     *  import { actorCalled } from '@serenity-js/core';
     *  import { BrowseTheWeb, Navigate } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('/search'),
     *      );
     *
     * @example <caption>Navigate to an absolute URL (overrides baseUrl)</caption>
     *  import { actorCalled } from '@serenity-js/core';
     *  import { BrowseTheWeb, Navigate } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('https://mycompany.org/login'),
     *      );
     *
     * @example <caption>Navigate to URL with timeout</caption>
     *  import { actorCalled, Duration } from '@serenity-js/core';
     *  import { BrowseTheWeb, Navigate } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('/search').withTimeout(Duration.ofSeconds(2)),
     *      );
     *
     * @param {Answerable<string>} url
     *  An absolute URL or path an {@link @serenity-js/core/lib/screenplay/actor~Actor} should navigate to
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction & { withTimeout: (duration: Answerable<Duration>) => Interaction }}
     *
     * @see {@link BrowseTheWeb}
     * @see {@link @serenity-js/core~Duration}
     */
    static to(url) {
        return new NavigateToUrl(url);
    }
    /**
     * @desc
     *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
     *  navigate back one page in the session history.
     *
     * @example <caption>Navigate to path relative to baseUrl</caption>
     *  import { actorCalled } from '@serenity-js/core';
     *  import { Ensure, endsWith } from '@serenity-js/assertions';
     *  import { BrowseTheWeb, Navigate } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('/first'),
     *          Navigate.to('/second'),
     *
     *          Navigate.back(),
     *
     *          Ensure.that(Website.url(), endsWith('/first')),
     *      );
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link BrowseTheWeb}
     * @see {@link @serenity-js/assertions~Ensure}
     * @see {@link @serenity-js/assertions/lib/expectations~endsWith}
     */
    static back() {
        return new NavigateBack();
    }
    /**
     * @desc
     *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
     *  navigate forward one page in the session history.
     *
     * @example <caption>Navigate to path relative to baseUrl</caption>
     *  import { actorCalled } from '@serenity-js/core';
     *  import { Ensure, endsWith } from '@serenity-js/assertions';
     *  import { BrowseTheWeb, Navigate } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('/first'),
     *          Navigate.to('/second'),
     *
     *          Navigate.back(),
     *          Navigate.forward(),
     *
     *          Ensure.that(Website.url(), endsWith('/second')),
     *      );
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link BrowseTheWeb}
     * @see {@link @serenity-js/assertions~Ensure}
     * @see {@link @serenity-js/assertions/lib/expectations~endsWith}
     */
    static forward() {
        return new NavigateForward();
    }
    /**
     * @desc
     *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor} to
     *  reload the current page.
     *
     * @example <caption>Navigate to path relative to baseUrl</caption>
     *  import { actorCalled } from '@serenity-js/core';
     *  import { Ensure, endsWith } from '@serenity-js/assertions';
     *  import { Navigate, BrowseTheWeb, DeleteCookies } from '@serenity-js/protractor';
     *
     *  actorCalled('Hannu')
     *      .whoCan(BrowseTheWeb.using(protractor.browser))
     *      .attemptsTo(
     *          Navigate.to('/login'),
     *          DeleteCookies.called('session_id'),
     *          Navigate.reloadPage(),
     *      );
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link BrowseTheWeb}
     * @see {@link DeleteCookies}
     * @see {@link @serenity-js/assertions~Ensure}
     * @see {@link @serenity-js/assertions/lib/expectations~endsWith}
     */
    static reloadPage() {
        return new ReloadPage();
    }
}
exports.Navigate = Navigate;
/**
 * @package
 */
class NavigateToUrl extends core_1.Interaction {
    constructor(url) {
        super();
        this.url = url;
    }
    /**
     * @desc
     *  Specifies timeout to wait for an Angular app to load.
     *  Please note that the timeout is ignored if you disable
     *  synchronisation with Angular.
     *
     * @param {Answerable<Duration>} duration
     */
    withTimeout(duration) {
        return new NavigateToUrlWithTimeout(this.url, duration);
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return actor.answer(this.url)
            .then(url => abilities_1.BrowseTheWeb.as(actor).get(url)
            .catch(error => {
            throw new core_1.TestCompromisedError(`Couldn't navigate to ${url}`, error);
        }));
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor navigates to ${this.url}`;
    }
}
/**
 * @package
 */
class NavigateToUrlWithTimeout extends core_1.Interaction {
    constructor(url, timeout) {
        super();
        this.url = url;
        this.timeout = timeout;
    }
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return Promise.all([
            actor.answer(this.url),
            actor.answer(this.timeout),
        ]).then(([url, timeout]) => abilities_1.BrowseTheWeb.as(actor).get(url, timeout.inMilliseconds()));
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor navigates to ${this.url} waiting up to ${this.timeout} for Angular to load`;
    }
}
/**
 * @package
 */
class NavigateBack extends core_1.Interaction {
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return promiseOf_1.promiseOf(abilities_1.BrowseTheWeb.as(actor).navigate().back());
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor navigates back in the browser history`;
    }
}
/**
 * @package
 */
class NavigateForward extends core_1.Interaction {
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return promiseOf_1.promiseOf(abilities_1.BrowseTheWeb.as(actor).navigate().forward());
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor navigates forward in the browser history`;
    }
}
/**
 * @package
 */
class ReloadPage extends core_1.Interaction {
    /**
     * @desc
     *  Makes the provided {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  perform this {@link @serenity-js/core/lib/screenplay~Interaction}.
     *
     * @param {UsesAbilities & AnswersQuestions} actor
     *  An {@link @serenity-js/core/lib/screenplay/actor~Actor} to perform this {@link @serenity-js/core/lib/screenplay~Interaction}
     *
     * @returns {PromiseLike<void>}
     *
     * @see {@link @serenity-js/core/lib/screenplay/actor~Actor}
     * @see {@link @serenity-js/core/lib/screenplay/actor~UsesAbilities}
     * @see {@link @serenity-js/core/lib/screenplay/actor~AnswersQuestions}
     */
    performAs(actor) {
        return promiseOf_1.promiseOf(abilities_1.BrowseTheWeb.as(actor).navigate().refresh());
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link @serenity-js/core/lib/screenplay~Activity}.
     *
     * @returns {string}
     */
    toString() {
        return io_1.formatted `#actor reloads the page`;
    }
}
