// @serenity-js/protractor/lib/screenplay/interactions/Switch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Switch = void 0;
const core_1 = require("@serenity-js/core");
const protractor_1 = require("protractor");
const abilities_1 = require("../abilities");
/**
 * @desc
 *  Instructs the {@link @serenity-js/core/lib/screenplay/actor~Actor}
 *  to switch to a different [frame](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/frame),
 *  [inline frame](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe),
 *  or browser window/tab.
 *
 * @example <caption>Lean Page Object describing a login form, embedded in an iframe</caption>
 *
 *  import { Target } from '@serenity-js/protractor';
 *  import { by } from 'protractor';
 *
 *  class LoginForm {
 *      static iframe           = Target.the('login form').located(by.tagName('iframe'));
 *      static usernameField    = Target.the('username field').located(by.css('[data-test="username"]'));
 *      static passwordField    = Target.the('password field').located(by.css('[data-test="password"]'));
 *      static submitButton     = Target.the('submit button').located(by.css(`button[type='submit']`));
 *  }
 *
 * @example <caption>Switch to an iframe and back</caption>
 *
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Switch, Enter, Click } from '@serenity-js/protractor';
 *  import { protractor } from 'protractor';
 *
 *  actorCalled('Francesca')
 *      .whoCan(BrowseTheWeb.using(protractor.browser))
 *      .attemptsTo(
 *          Switch.toFrame(LoginForm.iframe),
 *
 *          Enter.theValue('francesca@example.org').into(LoginForm.usernameField),
 *          Enter.theValue('correct-horse-battery-staple').into(LoginForm.passwordField),
 *          Click.on(LoginForm.submitButton),
 *
 *          Switch.toParentFrame(),
 *      );
 *
 * @example <caption>Perform activities in the context of an iframe</caption>
 *
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Switch, Enter, Click } from '@serenity-js/protractor';
 *  import { protractor } from 'protractor';
 *
 *  actorCalled('Francesca')
 *      .whoCan(BrowseTheWeb.using(protractor.browser))
 *      .attemptsTo(
 *          Switch.toFrame(LoginForm.iframe).and(
 *              Enter.theValue('francesca@example.org').into(LoginForm.usernameField),
 *              Enter.theValue('correct-horse-battery-staple').into(LoginForm.passwordField),
 *              Click.on(LoginForm.submitButton),
 *          ),
 *          // Note that Switch.toParentFrame() is invoked automatically
 *      );
 *
 * @example <caption>Switch to a new window/tab and back</caption>
 *
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Switch, Close } from '@serenity-js/protractor';
 *  import { protractor } from 'protractor';
 *
 *  actorCalled('Francesca')
 *      .whoCan(BrowseTheWeb.using(protractor.browser))
 *      .attemptsTo(
 *          Switch.toNewWindow(), // or: Switch.toWindow(...)
 *
 *          // perform some activities in the context of the new window
 *
 *          Close.currentWindow(),
 *
 *          Switch.toOriginalWindow(),
 *      );
 *
 * @example <caption>Perform activities in the context of a different window/tab</caption>
 *
 *  import { actorCalled } from '@serenity-js/core';
 *  import { BrowseTheWeb, Switch, Close } from '@serenity-js/protractor';
 *  import { protractor } from 'protractor';
 *
 *  actorCalled('Francesca')
 *      .whoCan(BrowseTheWeb.using(protractor.browser))
 *      .attemptsTo(
 *          Switch.toNewWindow().and(
 *              // perform some activities in the context of the new window
 *
 *              Close.currentWindow()
 *          ),
 *
 *          // Note that Switch.toOriginalWindow() is invoked automatically
 *      );
 *
 * @see {@link Close}
 * @see {@link BrowseTheWeb}
 */
class Switch {
    /**
     * @desc
     *  Switches the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to a [frame](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/frame)
     *  or an [inline frame](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)
     *  identified by its name, index or `Question<ElementFinder>`.
     *
     * @param {Answerable<ElementFinder | number | string>} targetOrIndex
     *
     * @returns {SwitchToFrame}
     *
     * @see {@link Switch.toParentFrame}
     * @see {@link Switch.toDefaultContent}
     * @see {@link Target}
     */
    static toFrame(targetOrIndex) {
        return new SwitchToFrame(targetOrIndex);
    }
    /**
     * @desc
     *  Sets the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to the parent of the current browsing context,
     *  i.e. an `iframe` in which the current `iframe` is nested.
     *
     *  If the current context is the top-level browsing context, the context remains unchanged.
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link Switch.toFrame}
     * @see https://w3c.github.io/webdriver/#switch-to-parent-frame
     * @see https://w3c.github.io/webdriver/#dfn-current-browsing-context
     */
    static toParentFrame() {
        return new SwitchToParentFrame();
    }
    /**
     * @desc
     *  Switches the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to the first frame on the page, or the main document
     *  when a page contains [`iframe`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)s.
     *
     * @returns {@serenity-js/core/lib/screenplay~Interaction}
     *
     * @see {@link Switch.toFrame}
     */
    static toDefaultContent() {
        return new SwitchToDefaultContent();
    }
    /**
     * @desc
     *  Switches the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to a browser tab/window identified by its
     *  name, index or [window handle](https://developer.mozilla.org/en-US/docs/Web/WebDriver/Commands/GetWindowHandles).
     *
     * @param {Answerable<string | number>} nameOrHandleOrIndex
     * @returns {SwitchToWindow}
     */
    static toWindow(nameOrHandleOrIndex) {
        return new SwitchToWindow(nameOrHandleOrIndex);
    }
    /**
     * @desc
     *  Switches the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to the most recently opened browser tab/window.
     *
     *  **Please note** that this behaviour might vary in some browsers if there are more than two windows opened
     *  at the same time.
     *
     * @returns {SwitchToNewWindow}
     */
    static toNewWindow() {
        return new SwitchToNewWindow();
    }
    /**
     * @desc
     *  Switches the current [browsing context](https://w3c.github.io/webdriver/#dfn-current-browsing-context)
     *  for future commands to the original window used when the {@link @serenity-js/core/lib/screenplay/actor~Actor}
     *  performed an interaction to {@link Navigate}.
     *
     *  **Please note** that this behaviour might vary in some browsers if there are more than two windows opened
     *  at the same time, as window handles might be ordered alphabetically instead of the order in which they were created.
     *
     * @returns {SwitchToOriginalWindow}
     */
    static toOriginalWindow() {
        return new SwitchToOriginalWindow();
    }
}
exports.Switch = Switch;
/**
 * @package
 */
class SwitchToFrame extends core_1.Interaction {
    constructor(targetOrIndex) {
        super();
        this.targetOrIndex = targetOrIndex;
    }
    and(...activities) {
        return new SwitchToFrameAndPerformActivities(this.targetOrIndex, activities);
    }
    performAs(actor) {
        return actor.answer(this.targetOrIndex)
            .then(targetOrIndex => abilities_1.BrowseTheWeb.as(actor).switchToFrame(targetOrIndex instanceof protractor_1.ElementFinder
            ? targetOrIndex.getWebElement() // https://github.com/angular/protractor/issues/1846#issuecomment-82634739
            : targetOrIndex));
    }
    toString() {
        return `#actor switches to frame: ${this.targetOrIndex}`;
    }
}
/**
 * @package
 */
class SwitchToFrameAndPerformActivities extends core_1.Task {
    constructor(targetOrIndex, activities) {
        super();
        this.targetOrIndex = targetOrIndex;
        this.activities = activities;
    }
    performAs(actor) {
        return actor.attemptsTo(new SwitchToFrame(this.targetOrIndex), ...this.activities, new SwitchToParentFrame());
    }
    toString() {
        return `#actor switches to frame: ${this.targetOrIndex}`;
    }
}
/**
 * @package
 */
class SwitchToParentFrame extends core_1.Interaction {
    performAs(actor) {
        return abilities_1.BrowseTheWeb.as(actor).switchToParentFrame();
    }
    toString() {
        return `#actor switches to parent frame`;
    }
}
/**
 * @package
 */
class SwitchToDefaultContent extends core_1.Interaction {
    performAs(actor) {
        return abilities_1.BrowseTheWeb.as(actor).switchToDefaultContent();
    }
    toString() {
        return `#actor switches to default content`;
    }
}
/**
 * @package
 */
class SwitchToWindow extends core_1.Interaction {
    constructor(nameOrHandleOrIndex) {
        super();
        this.nameOrHandleOrIndex = nameOrHandleOrIndex;
    }
    and(...activities) {
        return new SwitchToWindowAndPerformActivities(this.nameOrHandleOrIndex, activities);
    }
    performAs(actor) {
        return actor.answer(this.nameOrHandleOrIndex).then(index => abilities_1.BrowseTheWeb.as(actor).switchToWindow(index));
    }
    toString() {
        return `#actor switches to window: ${this.nameOrHandleOrIndex}`;
    }
}
/**
 * @package
 */
class SwitchToWindowAndPerformActivities extends core_1.Task {
    constructor(nameOrHandleOrIndex, activities) {
        super();
        this.nameOrHandleOrIndex = nameOrHandleOrIndex;
        this.activities = activities;
    }
    performAs(actor) {
        return abilities_1.BrowseTheWeb.as(actor).getCurrentWindowHandle()
            .then(currentWindow => {
            return actor.attemptsTo(new SwitchToWindow(this.nameOrHandleOrIndex), ...this.activities, new SwitchToWindow(currentWindow));
        });
    }
    toString() {
        return `#actor switches to window: ${this.nameOrHandleOrIndex}`;
    }
}
/**
 * @package
 */
class SwitchToNewWindow extends core_1.Interaction {
    constructor() {
        super();
    }
    and(...activities) {
        return new SwitchToNewWindowAndPerformActivities(activities);
    }
    performAs(actor) {
        return Promise.all([
            abilities_1.BrowseTheWeb.as(actor).getCurrentWindowHandle(),
            abilities_1.BrowseTheWeb.as(actor).getAllWindowHandles(),
        ])
            .then(([currentHandle, handles]) => handles.filter(handle => handle !== currentHandle))
            .then(handles => {
            if (handles.length === 0) {
                throw new core_1.LogicError(`No new window has been opened to switch to`);
            }
            return abilities_1.BrowseTheWeb.as(actor).switchToWindow(handles[handles.length - 1]);
        });
    }
    toString() {
        return `#actor switches to the new browser window`;
    }
}
/**
 * @package
 */
class SwitchToNewWindowAndPerformActivities extends core_1.Task {
    constructor(activities) {
        super();
        this.activities = activities;
    }
    performAs(actor) {
        return abilities_1.BrowseTheWeb.as(actor).getCurrentWindowHandle()
            .then(currentWindow => {
            return actor.attemptsTo(new SwitchToNewWindow(), ...this.activities, new SwitchToWindow(currentWindow));
        });
    }
    toString() {
        return `#actor switches to the new window`;
    }
}
/**
 * @package
 */
class SwitchToOriginalWindow extends core_1.Interaction {
    constructor() {
        super();
    }
    performAs(actor) {
        return abilities_1.BrowseTheWeb.as(actor).switchToOriginalWindow();
    }
    toString() {
        return `#actor switches back to the original browser window`;
    }
}
