// @serenity-js/core/lib/screenplay/Interaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Interaction = void 0;
/**
 * @desc
 *  Directly exercises the {@link Actor}'s {@link Ability} to interact
 *  with the System Under Test.
 *
 * @implements {Activity}
 * @see {@link Ability}
 * @see {@link Actor}
 */
class Interaction {
    static where(description, interaction) {
        return new DynamicallyGeneratedInteraction(description, interaction);
    }
}
exports.Interaction = Interaction;
/**
 * @package
 */
class DynamicallyGeneratedInteraction extends Interaction {
    constructor(description, interaction) {
        super();
        this.description = description;
        this.interaction = interaction;
    }
    /**
     * @desc
     *  Makes the provided {@link Actor}
     *  perform this {@link Interaction}.
     *
     * @param {Actor} actor
     * @returns {Promise<void>}
     *
     * @see {@link Actor}
     */
    performAs(actor) {
        try {
            return Promise.resolve(this.interaction(actor));
        }
        catch (error) {
            return Promise.reject(error);
        }
    }
    /**
     * @desc
     *  Generates a description to be used when reporting this {@link Activity}.
     *
     * @returns {string}
     */
    toString() {
        return this.description;
    }
}
