// @serenity-js/core/lib/io/Version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Version = void 0;
const semver = require("semver");
const tiny_types_1 = require("tiny-types");
/**
 * @desc
 *  A tiny type describing a version number, like `1.2.3`
 *
 * @extends {tiny-types~TinyType}
 */
class Version extends tiny_types_1.TinyType {
    /**
     * @param {string} version
     */
    constructor(version) {
        super();
        this.version = version;
        tiny_types_1.ensure('version', version, tiny_types_1.isDefined(), tiny_types_1.isString(), isValid());
    }
    /**
     * @param {string} version
     * @returns {Version}
     */
    static fromJSON(version) {
        return new Version(version);
    }
    /**
     * @param {Version} other
     * @returns {boolean}
     */
    isAtLeast(other) {
        return semver.gte(this.version, other.version);
    }
    /**
     * @returns {number}
     *  Major version number of a given package version, i.e. `1` in `1.2.3`
     */
    major() {
        return Number(this.version.split('.')[0]);
    }
    /**
     * @returns {string}
     */
    toString() {
        return `${this.version}`;
    }
}
exports.Version = Version;
/**
 * @package
 */
function isValid() {
    return tiny_types_1.Predicate.to(`be a valid version number`, (version) => !!semver.valid(version));
}
