// @serenity-js/core/lib/events/ActivityRelatedArtifactArchived.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityRelatedArtifactArchived = void 0;
const tiny_types_1 = require("tiny-types");
const io_1 = require("../io");
const model_1 = require("../model");
const ArtifactArchived_1 = require("./ArtifactArchived");
// todo: there should be two types of events here:
//  - ActivityArtifactGenerated
//  - SceneArtifactGenerated
class ActivityRelatedArtifactArchived extends ArtifactArchived_1.ArtifactArchived {
    constructor(sceneId, activityId, name, type, path, timestamp) {
        super(sceneId, name, type, path, timestamp);
        this.activityId = activityId;
        tiny_types_1.ensure('activityId', activityId, tiny_types_1.isDefined());
    }
    static fromJSON(o) {
        return new ActivityRelatedArtifactArchived(model_1.CorrelationId.fromJSON(o.sceneId), model_1.CorrelationId.fromJSON(o.activityId), model_1.Name.fromJSON(o.name), model_1.Artifact.ofType(o.type), io_1.Path.fromJSON(o.path), model_1.Timestamp.fromJSON(o.timestamp));
    }
    toJSON() {
        return {
            sceneId: this.sceneId.toJSON(),
            activityId: this.activityId.toJSON(),
            name: this.name.toJSON(),
            type: this.type.name,
            path: this.path.toJSON(),
            timestamp: this.timestamp.toJSON(),
        };
    }
}
exports.ActivityRelatedArtifactArchived = ActivityRelatedArtifactArchived;
