// @serenity-js/core/lib/events/AsyncOperationFailed.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncOperationFailed = void 0;
const tiny_types_1 = require("tiny-types");
const io_1 = require("../io");
const model_1 = require("../model");
const DomainEvent_1 = require("./DomainEvent");
class AsyncOperationFailed extends DomainEvent_1.DomainEvent {
    constructor(error, correlationId, timestamp) {
        super(timestamp);
        this.error = error;
        this.correlationId = correlationId;
        tiny_types_1.ensure('error', error, tiny_types_1.isDefined());
        tiny_types_1.ensure('correlationId', correlationId, tiny_types_1.isDefined());
    }
    static fromJSON(o) {
        return new AsyncOperationFailed(io_1.ErrorSerialiser.deserialise(o.error), model_1.CorrelationId.fromJSON(o.correlationId), model_1.Timestamp.fromJSON(o.timestamp));
    }
    toJSON() {
        return {
            correlationId: this.correlationId.toJSON(),
            error: io_1.ErrorSerialiser.serialise(this.error),
            timestamp: this.timestamp.toJSON(),
        };
    }
}
exports.AsyncOperationFailed = AsyncOperationFailed;
